### PR TITLE
phoenix-pilot: new mathematical model (attitude, 1/3)

### PR DESCRIPTION
**Pull request by niewim19**
## Description

New mathematical model is introduced to phoenix-pilot. First stage implements only attitude estimation.
 - added control vector
 - using measurements of gyroscope in prediction step for integration purposes without unnecessarily storing them in the state vector  

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: zturn drone.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: ekf/kalman_implem.c [hunk 1]
```diff
@@ -111,200 +97,193 @@ void kmn_configRead(kalman_init_t *initVals)
...
+	const quat_t qState = {.a = kmn_vecAt(state, QA), .i = kmn_vecAt(state, QB), .j = kmn_vecAt(state, QC), .k = kmn_vecAt(state, QD)};
+	const quat_t bwState = {.a = 0, .i = kmn_vecAt(state, BWX), .j = kmn_vecAt(state, BWY), .k = kmn_vecAt(state, BWZ)};
+
+	/* quaternionized angular rate from U vector */
+	quat_t wMeas = {.a = 0, .i = kmn_vecAt(U, UWX), .j = kmn_vecAt(U, UWY), .k = kmn_vecAt(U, UWZ)};
+
+	/* quaternionized angular rate and rotation quaternion estimates */
+	quat_t qEst, qtmp;
+
+	float dt = (float)timeStep / 1000000.f;
```
Comment (kemonats) [inline ekf/kalman_implem.c:128]:
```suggestion
	const float dt = (float)timeStep / 1000000.f;
```

File: ekf/kalman_implem.c [hunk 2]
```diff
@@ -111,200 +97,193 @@ void kmn_configRead(kalman_init_t *initVals)
...
+{
+	const quat_t qState = {.a = kmn_vecAt(state, QA), .i = kmn_vecAt(state, QB), .j = kmn_vecAt(state, QC), .k = kmn_vecAt(state, QD)};
+	const quat_t bwState = {.a = 0, .i = kmn_vecAt(state, BWX), .j = kmn_vecAt(state, BWY), .k = kmn_vecAt(state, BWZ)};
+	const quat_t wMeas = {.a = 0, .i = kmn_vecAt(U, UWX), .j = kmn_vecAt(U, UWY), .k = kmn_vecAt(U, UWZ)};
+
+	const float dt = (float)timeStep / 1000000;
+
+	/* d(f_q)/d(q) variables */
+	float dfqdqData[4 * 4];
+	matrix_t dfqdq = { .data = dfqdqData, .rows = 3, .cols = 3, .transposed = 0 };
```
Comment (kemonats) [inline ekf/kalman_implem.c:199]: 4 x 4 or 3 x 3

File: ekf/kalman_implem.h [hunk 3]
```diff
@@ -155,6 +91,12 @@ int verbose;
 
 extern void kmn_configRead(kalman_init_t *initVals);
 
+/* Reads from a matrix like from a untransposed column vector directly */
+static inline float kmn_vecAt(const matrix_t *M, unsigned int i)
+{
+	return M->data[i];
```
Comment (kemonats) [inline ekf/kalman_implem.h:97]: What if index `i` will be outside the range of the array.

Comment (niewim19) [reply ekf/kalman_implem.h]: Implemented simple inline check. Thanks!

File: ekf/kalman_implem.c (hunk 2 above)
Comment (niewim19) [reply ekf/kalman_implem.c]: Wow... very good catch! Thank you very much!

It should be 4x4 as it is derivative of quaternion with respect to quaternion (H^4 over H^4).

File: ekf/kalman_implem.c (hunk 1 above)
Comment (niewim19) [reply ekf/kalman_implem.c]: Implemented, thanks!

**Comment by niewim19** [comment]
This PR is being closed and commits introduced in it are merged in a cumulative PR:
 - https://github.com/phoenix-pilot/phoenix-pilot-core/pull/199

**Pull request closed without merging**